### PR TITLE
Remove `init` from pageserver CLI

### PR DESCRIPTION
**Pull request by SomeoneToIgnore**
Removes pageserver `init` CLI command.

* `pageserver.toml` config file is now only created or updated when pageserver `start` command gets `--update-config` flag.
`id` is still a protected field that is not allowed to be updated in the existing config file

* initial tenant and timeline creation is now only done in `neon_local init`, for that entire pageserver is started to serve two HTTP creation commands and then stopped.

Cloud does not use initial timelines at all, Python tests continue using the same "initial timeline" implicitly brought by the `neon_local init` invocation

### CODE REVIEW

File: control_plane/src/local_env.rs [hunk 1]
```diff
@@ -320,7 +320,7 @@ impl LocalEnv {
 
         if !repopath.exists() {
             bail!(
-                "Zenith config is not found in {}. You need to run 'zenith init' first",
+                "Zenith config is not found in {}. You need to run 'neon_local init' first",
```
Comment (hlinnaka) [inline control_plane/src/local_env.rs:323]: while we're at it:
```suggestion
                "Neon config is not found in {}. You need to run 'neon_local init' first",
```

File: control_plane/src/storage.rs [hunk 2]
```diff
@@ -63,7 +63,7 @@ impl ResponseErrorMessageExt for Response {
         Err(PageserverHttpError::Response(
             match self.json::<HttpErrorBody>() {
                 Ok(err_body) => format!("Error: {}", err_body.msg),
-                Err(_) => format!("Http error ({}) at {}.", status.as_u16(), url),
+                Err(_) => format!("Http error ({}) at {url}.", status.as_u16()),
```
Comment (hlinnaka) [inline control_plane/src/storage.rs:66]: I find it confusing to mix the `".. {}", var` and `"{var}"` styles in the same message, so I'd prefer to leave it as it is.

File: control_plane/src/storage.rs [hunk 3]
```diff
@@ -211,15 +192,35 @@ impl PageServerNode {
...
+        self.start_node(config_overrides, &self.repo_path(), false)
+    }
+
+    fn start_node(
+        &self,
+        config_overrides: &[&str],
+        datadir: &Path,
+        update_config: bool,
+    ) -> anyhow::Result<()> {
+        println!(
```
Comment (hlinnaka) [inline control_plane/src/storage.rs:204]: Note that the old code used `print!()` rather than `println!()`, so this will print an extra empty line

File: control_plane/src/storage.rs [hunk 4]
```diff
@@ -500,7 +491,7 @@ impl PageServerNode {
         let timeline_infos: Vec<TimelineInfo> = self
             .http_request(
                 Method::GET,
-                format!("{}/tenant/{}/timeline", self.http_base_url, tenant_id),
+                format!("{}/tenant/{tenant_id}/timeline", self.http_base_url),
```
Comment (hlinnaka) [inline control_plane/src/storage.rs:494]: same as above, I'd prefer not mixing the format arg styles on same line

File: control_plane/src/storage.rs [hunk 5]
```diff
@@ -515,22 +506,27 @@ impl PageServerNode {
         new_timeline_id: Option<ZTimelineId>,
         ancestor_start_lsn: Option<Lsn>,
         ancestor_timeline_id: Option<ZTimelineId>,
-    ) -> anyhow::Result<Option<TimelineInfo>> {
-        let timeline_info_response = self
-            .http_request(
-                Method::POST,
-                format!("{}/tenant/{}/timeline", self.http_base_url, tenant_id),
+    ) -> anyhow::Result<TimelineInfo> {
```
Comment (hlinnaka) [inline control_plane/src/storage.rs:509]: Why is this change needed?

**Review by hlinnaka — COMMENTED** [review]
Ah, that's a patch to my liking :smile: :
```
 9 files changed, 218 insertions(+), 340 deletions(-)
```
Does starting the `pageserver` binary now implicitly create the target directory if it doesn't already exist? If so, that seems a bit fragile, I think I'd prefer to get an error, if for example the repository directory doesn't exist because you forgot to mount the filesystem it's in.

File: control_plane/src/storage.rs (hunk 3 above)
Comment (SomeoneToIgnore) [reply control_plane/src/storage.rs:204]: Yes, yet I've found that more natural when dealing with failing pageserver: 

with empty line:
```
Work/neon/neon kb/pageserver-no-init:origin/kb/pageserver-no-init*​ ⇡❯ rm -rf .neon/ && ./target/debug/neon_local init
Starting pageserver at '127.0.0.1:64000' in '.neon'
thread 'main' panicked at '((((((((((((((((', pageserver/src/bin/pageserver.rs:41:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/panicking.rs:142:14
   2: pageserver::main
             at ./pageserver/src/bin/pageserver.rs:41:9
   3: core::ops::function::FnOnce::call_once
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/ops/function.rs:248:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
pageserver init failed: Pageserver failed to start. See console output and '.neon/pageserver.log' for details.
```

without:
```
Work/neon/neon kb/pageserver-no-init:origin/kb/pageserver-no-init*​ ⇡❯ rm -rf .neon/ && ./target/debug/neon_local init
Starting pageserver at '127.0.0.1:64000' in '.neon'thread 'main' panicked at '((((((((((((((((', pageserver/src/bin/pageserver.rs:41:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/panicking.rs:142:14
   2: pageserver::main
             at ./pageserver/src/bin/pageserver.rs:41:9
   3: core::ops::function::FnOnce::call_once
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/ops/function.rs:248:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
pageserver init failed: Pageserver failed to start. See console output and '.neon/pageserver.log' for details.
```

File: control_plane/src/storage.rs (hunk 5 above)
Comment (SomeoneToIgnore) [reply control_plane/src/storage.rs:509]: `Option<TimelineInfo>` is `None` according to our logic only if we fail to create a timeline:
https://github.com/neondatabase/neon/blob/e593cbaabafafb6f54c9ddcd5d1d9f04d1bd4490/pageserver/src/http/routes.rs#L244

So instead of doing various `.unwrap()` or other error handling above the client, I've simplified it to validating the presence here.

**Comment by SomeoneToIgnore** [comment]
> Does starting the pageserver binary now implicitly create the target directory if it doesn't already exist?

The `tenants/` one? Yes, now we do crate it implicitly. Do you think that base path should be mounted instead anyway? We'll fail if that's not done, so either I don't understand it or it should be ok.

We could somehow alter `--update-config` to something more generic and allow dir creation along with that operation too, yet it slightly becomes to resemble `--init` flag now 🙂

**Comment by SomeoneToIgnore** [comment]
A PR to fix e2e tests: https://github.com/neondatabase/cloud/pull/2058
Other tests seem to flack for known, daemonize-related issues.

**Comment by hlinnaka** [comment]
>> Does starting the pageserver binary now implicitly create the target directory if it doesn't already exist?
> 
> The `tenants/` one? Yes, now we do crate it implicitly. Do you think that base path should be mounted instead anyway? We'll fail if that's not done, so either I don't understand it or it should be ok.

I meant the directory that contains `tenants/`. The data directory looks like this:

```
<path>/tenants/
<path>/pageserver.log
<path>/pageserver.pid
<path>/pageserver.toml
```

The log file is created on startup. The pid-file is also created. The config file isn't mandatory I guess. But the question is, if `<path>` itself doesn't exist, do we create it automatically, or error out? And if `<path>/tenants` doesn't already exist, do we create it or error out?

Imagine that pageserver runs on a host with two filesystems. The root filesystem, and 'data' that holds all the pageserver data. At server startup, we do this:
1. `mount /data`
2. `pageserver --workdir /data/pageserver-data`

What happens if the `mount /data` command fails? Does pageserver still start up and automatically create a new empty data directory as `/data/pageserver-data/`, now on the root filesystem? Or does it refuse to start? I think refusing to start is better.

> We could somehow alter `--update-config` to something more generic and allow dir creation along with that operation too, yet it slightly becomes to resemble `--init` flag now slightly_smiling_face

Yeah, I think an explicit `init` option still makes sense. But without creating the initial tenant and timeline.

**Comment by SomeoneToIgnore** [comment]
> I meant the directory that contains tenants/. The data directory looks like this:

Now I understand.
This is a workdir, the one that's being checked for existence implicitly, with `.canonicalize`:
https://github.com/neondatabase/neon/blob/d29c545b5d214413e5fc1199b2aece650561cc44/pageserver/src/bin/pageserver.rs#L107-L111

If we try to canonicalize or do `env::set_current_dir` (we do that later below), we fail for non-existent path.
So we don't create the workdir, rather implicitly failing.

This means I can avoid changing the code to be `--init`, after all.

**Review by LizardWizzard — APPROVED** [review]
Looks good to me

I’m for avoiding init and bailing when passed directory doesn’t exist. Though I’m not sure that this is followed by safekeepers, need to check

UPD: safekeeper will start even without passed datadir argument. It will fallback to current working directory. It is the simplest behavior, but if we want to guard against such config errors we better use the same behavior everywhere. Thoughts?

File: control_plane/src/storage.rs [hunk 6]
```diff
@@ -139,67 +135,52 @@ impl PageServerNode {
                 .collect::<Vec<_>>()
                 .join(",")
         );
-        let mut args = Vec::with_capacity(20);
-
-        args.push("--init");
```
Comment (ololobus) [inline control_plane/src/storage.rs:144]: Probably you will need to modify this script too
https://github.com/neondatabase/neon/blob/dc102197df0469da544463253e636395d1d33789/.github/ansible/scripts/init_pageserver.sh#L28-L30

Comment (SomeoneToIgnore) [reply control_plane/src/storage.rs:143]: Thank you, I've considered options and decided to replicate `--init` from safekeeper as discussed above.

Comment (SomeoneToIgnore) [reply control_plane/src/storage.rs:143]: Added [a test](https://github.com/neondatabase/neon/pull/2272/files#diff-85440eac3bb111914ec6d1e6029a82c37f6dc7f303dda8a33d9c7d56e4ffc8b9R22) for the new `--init` command to replicate this script.
